README
Task::BeLike::HMA
=================

Task::BeLike::HMA is HMA's personal Task to install the Perl
modules he likes to have available.

See "Makefile.PL" for details.


SYNOPSIS / INSTALLATION

    perl Makefile.PL
    make

    # re-run Makefile.PL to see if all modules got installed
    perl Makefile.PL

    # make test and install if you want (not required)
    make test
    make install


DEPENDENCIES

This Task checks if recent versions of Perl toolchain modules are available.
If not, installation of Task::BeLike::HMA::Toolchain is suggested and nothing
is installed.

See https://github.com/hma/Task-BeLike-HMA-Toolchain

The required minimum versions of toolchain modules are:

    CPAN 1.9301 (in Perl 5.8.9 core)
    Module::Build 0.340201 (in 5.10.1 core)


COPYRIGHT AND LICENSE

Copyright (c) 2010-2011 Henning Manske. All rights reserved.

This module is free software. You can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/

This module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Task-BeLike-HMA

0.09_01  Aug 01, 2011

        - Added modules
            * CPAN::Mini
        - Moved dependencies to Task-BeLike-HMA-Toolchain
            * Archive::Zip
            * Module::CoreList
        - Raise dependency on Module::Build from 0.340201 to 0.3603

0.09    Jul 23, 2011

        - Added modules
            * App::cpanlistchanges
            * App::cpanminus
            * App::cpanoutdated
            * App::moduleshere
            * App::pmuninstall
            * local::lib
        - Reflect changes in LWP 6 to get SSL support
        - Test::Memory::Cycle requires Perl 5.8.2
        - Workaround for Module::CPANTS::Analyse requiring
          Test::YAML::Meta::Version (RT #65903)
        - Adjust required versions for many dependencies
        - Add more Perl version checks for legacy Perls
        - Add Task::BeLike::HMA::Toolchain to metadata
        - Convert README to Markdown
        - Upgrade to Module::Install 1.01

0.08    Dec 15, 2010

        - Added modules
            * Carp::Always
            * Crypt::SSLeay
            * Data::Dump
        - Update Perl::Critic if required
        - Switch GitHub URLs to https

0.06    Oct 24, 2010

        - FIX: Re-enforced usage of CPAN.pm for installation

0.05    Oct 24, 2010

        - Module::Version requires Getopt::Long 2.36
        - Added modules
            * aliased
            * Carp::Clan
            * Config::General
            * Module::Install
            * Path::Class
            * Test::Memory::Cycle
            * Test::Requires

0.03    Oct 10, 2010

        - Made dependencies I use explicit
            * Archive::Zip
            * Parse::RecDescent
            * Sub::Exporter
        - Moved dependencies to Task-BeLike-HMA-Toolchain
            * Digest::MD5
            * File::Spec
            * Storable
            * Test::More
            * version.pm
        - Changed minimum Perl version to 5.6.1
        - Force Module::AutoInstall to use CPAN.pm
        - Installation order adjusted

0.01    Sep 12, 2010

        - Initial commit and release to GitHub


inc/Module/Install.pm
#line 1
package Module::Install;

# For any maintainers:
# The load order for Module::Install is a bit magic.
# It goes something like this...
#
# IF ( host has Module::Install installed, creating author mode ) {
#     1. Makefile.PL calls "use inc::Module::Install"
#     2. $INC{inc/Module/Install.pm} set to installed version of inc::Module::Install
#     3. The installed version of inc::Module::Install loads
#     4. inc::Module::Install calls "require Module::Install"
#     5. The ./inc/ version of Module::Install loads
# } ELSE {
#     1. Makefile.PL calls "use inc::Module::Install"
#     2. $INC{inc/Module/Install.pm} set to ./inc/ version of Module::Install
#     3. The ./inc/ version of Module::Install loads
# }

use 5.005;
use strict 'vars';
use Cwd        ();
use File::Find ();
use File::Path ();

use vars qw{$VERSION $MAIN};
BEGIN {
	# All Module::Install core packages now require synchronised versions.
	# This will be used to ensure we don't accidentally load old or
	# different versions of modules.
	# This is not enforced yet, but will be some time in the next few
	# releases once we can make sure it won't clash with custom
	# Module::Install extensions.
	$VERSION = '0.9999';

	# Storage for the pseudo-singleton
	$MAIN    = undef;

	*inc::Module::Install::VERSION = *VERSION;
	@inc::Module::Install::ISA     = __PACKAGE__;

}

sub import {
	my $class = shift;
	my $self  = $class->new(@_);
	my $who   = $self->_caller;

	#-------------------------------------------------------------
	# all of the following checks should be included in import(),
	# to allow "eval 'require Module::Install; 1' to test
	# installation of Module::Install. (RT #51267)
	#-------------------------------------------------------------

	# Whether or not inc::Module::Install is actually loaded, the
	# $INC{inc/Module/Install.pm} is what will still get set as long as
	# the caller loaded module this in the documented manner.
	# If not set, the caller may NOT have loaded the bundled version, and thus
	# they may not have a MI version that works with the Makefile.PL. This would
	# result in false errors or unexpected behaviour. And we don't want that.
	my $file = join( '/', 'inc', split /::/, __PACKAGE__ ) . '.pm';
	unless ( $INC{$file} ) { die <<"END_DIE" }

Please invoke ${\__PACKAGE__} with:

	use inc::${\__PACKAGE__};

not:

	use ${\__PACKAGE__};

END_DIE

	# This reportedly fixes a rare Win32 UTC file time issue, but
	# as this is a non-cross-platform XS module not in the core,
	# we shouldn't really depend on it. See RT #24194 for detail.
	# (Also, this module only supports Perl 5.6 and above).
	eval "use Win32::UTCFileTime" if $^O eq 'MSWin32' && $] >= 5.006;

	# If the script that is loading Module::Install is from the future,
	# then make will detect this and cause it to re-run over and over
	# again. This is bad. Rather than taking action to touch it (which
	# is unreliable on some platforms and requires write permissions)
	# for now we should catch this and refuse to run.
	if ( -f $0 ) {
		my $s = (stat($0))[9];

		# If the modification time is only slightly in the future,
		# sleep briefly to remove the problem.
		my $a = $s - time;
		if ( $a > 0 and $a < 5 ) { sleep 5 }

		# Too far in the future, throw an error.
		my $t = time;
		if ( $s > $t ) { die <<"END_DIE" }

Your installer $0 has a modification time in the future ($s > $t).

This is known to create infinite loops in make.

Please correct this, then run $0 again.

END_DIE
	}


	# Build.PL was formerly supported, but no longer is due to excessive
	# difficulty in implementing every single feature twice.
	if ( $0 =~ /Build.PL$/i ) { die <<"END_DIE" }

Module::Install no longer supports Build.PL.

It was impossible to maintain duel backends, and has been deprecated.

Please remove all Build.PL files and only use the Makefile.PL installer.

END_DIE

	#-------------------------------------------------------------

	# To save some more typing in Module::Install installers, every...
	# use inc::Module::Install
	# ...also acts as an implicit use strict.
	$^H |= strict::bits(qw(refs subs vars));

	#-------------------------------------------------------------

	unless ( -f $self->{file} ) {
		foreach my $key (keys %INC) {
			delete $INC{$key} if $key =~ /Module\/Install/;
		}

		local $^W;
		require "$self->{path}/$self->{dispatch}.pm";
		File::Path::mkpath("$self->{prefix}/$self->{author}");
		$self->{admin} = "$self->{name}::$self->{dispatch}"->new( _top => $self );
		$self->{admin}->init;
		@_ = ($class, _self => $self);
		goto &{"$self->{name}::import"};
	}

	local $^W;
	*{"${who}::AUTOLOAD"} = $self->autoload;
	$self->preload;

	# Unregister loader and worker packages so subdirs can use them again
	delete $INC{'inc/Module/Install.pm'};
	delete $INC{'Module/Install.pm'};

	# Save to the singleton
	$MAIN = $self;

	return 1;
}

sub autoload {
	my $self = shift;
	my $who  = $self->_caller;
	my $cwd  = Cwd::cwd();
	my $sym  = "${who}::AUTOLOAD";
	$sym->{$cwd} = sub {
		my $pwd = Cwd::cwd();
		if ( my $code = $sym->{$pwd} ) {
			# Delegate back to parent dirs
			goto &$code unless $cwd eq $pwd;
		}
		unless ($$sym =~ s/([^:]+)$//) {
			# XXX: it looks like we can't retrieve the missing function
			# via $$sym (usually $main::AUTOLOAD) in this case.
			# I'm still wondering if we should slurp Makefile.PL to
			# get some context or not ...
			my ($package, $file, $line) = caller;
			die <<"EOT";
Unknown function is found at $file line $line.
Execution of $file aborted due to runtime errors.

If you're a contributor to a project, you may need to install
some Module::Install extensions from CPAN (or other repository).
If you're a user of a module, please contact the author.
EOT
		}
		my $method = $1;
		if ( uc($method) eq $method ) {
			# Do nothing
			return;
		} elsif ( $method =~ /^_/ and $self->can($method) ) {
			# Dispatch to the root M:I class
			return $self->$method(@_);
		}

		# Dispatch to the appropriate plugin
		unshift @_, ( $self, $1 );
		goto &{$self->can('call')};
	};
}

sub preload {
	my $self = shift;
	unless ( $self->{extensions} ) {
		$self->load_extensions(
			"$self->{prefix}/$self->{path}", $self
		);
	}

	my @exts = @{$self->{extensions}};
	unless ( @exts ) {
		@exts = $self->{admin}->load_all_extensions;
	}

	my %seen;
	foreach my $obj ( @exts ) {
		while (my ($method, $glob) = each %{ref($obj) . '::'}) {
			next unless $obj->can($method);
			next if $method =~ /^_/;
			next if $method eq uc($method);
			$seen{$method}++;
		}
	}

	my $who = $self->_caller;
	foreach my $name ( sort keys %seen ) {
		local $^W;
		*{"${who}::$name"} = sub {
			${"${who}::AUTOLOAD"} = "${who}::$name";
			goto &{"${who}::AUTOLOAD"};
		};
	}
}

sub new {
	my ($class, %args) = @_;

	delete $INC{'FindBin.pm'};
	{
		# to suppress the redefine warning
		local $SIG{__WARN__} = sub {};
		require FindBin;
	}

	# ignore the prefix on extension modules built from top level.
	my $base_path = Cwd::abs_path($FindBin::Bin);
	unless ( Cwd::abs_path(Cwd::cwd()) eq $base_path ) {
		delete $args{prefix};
	}
	return $args{_self} if $args{_self};

	$args{dispatch} ||= 'Admin';
	$args{prefix}   ||= 'inc';
	$args{author}   ||= ($^O eq 'VMS' ? '_author' : '.author');
	$args{bundle}   ||= 'inc/BUNDLES';
	$args{base}     ||= $base_path;
	$class =~ s/^\Q$args{prefix}\E:://;
	$args{name}     ||= $class;
	$args{version}  ||= $class->VERSION;
	unless ( $args{path} ) {
		$args{path}  = $args{name};
		$args{path}  =~ s!::!/!g;
	}
	$args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";
	$args{wrote}      = 0;

	bless( \%args, $class );
}

sub call {
	my ($self, $method) = @_;
	my $obj = $self->load($method) or return;
        splice(@_, 0, 2, $obj);
	goto &{$obj->can($method)};
}

sub load {
	my ($self, $method) = @_;

	$self->load_extensions(
		"$self->{prefix}/$self->{path}", $self
	) unless $self->{extensions};

	foreach my $obj (@{$self->{extensions}}) {
		return $obj if $obj->can($method);
	}

	my $admin = $self->{admin} or die <<"END_DIE";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END_DIE

	my $obj = $admin->load($method, 1);
	push @{$self->{extensions}}, $obj;

	$obj;
}

sub load_extensions {
	my ($self, $path, $top) = @_;

	my $should_reload = 0;
	unless ( grep { ! ref $_ and lc $_ eq lc $self->{prefix} } @INC ) {
		unshift @INC, $self->{prefix};
		$should_reload = 1;
	}

	foreach my $rv ( $self->find_extensions($path) ) {
		my ($file, $pkg) = @{$rv};
		next if $self->{pathnames}{$pkg};

		local $@;
		my $new = eval { local $^W; require $file; $pkg->can('new') };
		unless ( $new ) {
			warn $@ if $@;
			next;
		}
		$self->{pathnames}{$pkg} =
			$should_reload ? delete $INC{$file} : $INC{$file};
		push @{$self->{extensions}}, &{$new}($pkg, _top => $top );
	}

	$self->{extensions} ||= [];
}

sub find_extensions {
	my ($self, $path) = @_;

	my @found;
	File::Find::find( sub {
		my $file = $File::Find::name;
		return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
		my $subpath = $1;
		return if lc($subpath) eq lc($self->{dispatch});

		$file = "$self->{path}/$subpath.pm";
		my $pkg = "$self->{name}::$subpath";
		$pkg =~ s!/!::!g;

		# If we have a mixed-case package name, assume case has been preserved
		# correctly.  Otherwise, root through the file to locate the case-preserved
		# version of the package name.
		if ( $subpath eq lc($subpath) || $subpath eq uc($subpath) ) {
			my $content = Module::Install::_read($subpath . '.pm');
			my $in_pod  = 0;
			foreach ( split //, $content ) {
				$in_pod = 1 if /^=\w/;
				$in_pod = 0 if /^=cut/;
				next if ($in_pod || /^=cut/);  # skip pod text
				next if /^\s*#/;               # and comments
				if ( m/^\s*package\s+($pkg)\s*;/i ) {
					$pkg = $1;
					last;
				}
			}
		}

		push @found, [ $file, $pkg ];
	}, $path ) if -d $path;

	@found;
}





#####################################################################
# Common Utility Functions

sub _caller {
	my $depth = 0;
	my $call  = caller($depth);
	while ( $call eq __PACKAGE__ ) {
		$depth++;
		$call = caller($depth);
	}
	return $call;
}

# Done in evals to avoid confusing Perl::MinimumVersion
eval( $] >= 5.006 ? <<'END_NEW' : <<'END_OLD' ); die $@ if $@;
sub _read {
	local *FH;
	open( FH, '<', $_[0] ) or die "open($_[0]): $!";
	my $string = do { local $/; <FH> };
	close FH or die "close($_[0]): $!";
	return $string;
}
END_NEW
sub _read {
	local *FH;
	open( FH, "< $_[0]"  ) or die "open($_[0]): $!";
	my $string = do { local $/; <FH> };
	close FH or die "close($_[0]): $!";
	return $string;
}
END_OLD

sub _readperl {
	my $string = Module::Install::_read($_[0]);
	$string =~ s/(?:\015{1,2}\012|\015|\012)/\n/sg;
	$string =~ s/(\n)\n*__(?:DATA|END)__\b.*\z/$1/s;
	$string =~ s/\n\n=\w+.+?\n\n=cut\b.+?\n+/\n\n/sg;
	return $string;
}

sub _readpod {
	my $string = Module::Install::_read($_[0]);
	$string =~ s/(?:\015{1,2}\012|\015|\012)/\n/sg;
	return $string if $_[0] =~ /\.pod\z/;
	$string =~ s/(^|\n=cut\b.+?\n+)[^=\s].+?\n(\n=\w+|\z)/$1$2/sg;
	$string =~ s/\n*=pod\b[^\n]*\n+/\n\n/sg;
	$string =~ s/\n*=cut\b[^\n]*\n+/\n\n/sg;
	$string =~ s/^\n+//s;
	return $string;
}

# Done in evals to avoid confusing Perl::MinimumVersion
eval( $] >= 5.006 ? <<'END_NEW' : <<'END_OLD' ); die $@ if $@;
sub _write {
	local *FH;
	open( FH, '>', $_[0] ) or die "open($_[0]): $!";
	foreach ( 1 .. $#_ ) {
		print FH $_[$_] or die "print($_[0]): $!";
	}
	close FH or die "close($_[0]): $!";
}
END_NEW
sub _write {
	local *FH;
	open( FH, "> $_[0]"  ) or die "open($_[0]): $!";
	foreach ( 1 .. $#_ ) {
		print FH $_[$_] or die "print($_[0]): $!";
	}
	close FH or die "close($_[0]): $!";
}
END_OLD

# _version is for processing module versions (eg, 1.03_05) not
# Perl versions (eg, 5.8.1).
sub _version ($) {
	my $s = shift || 0;
	my $d =()= $s =~ /(\.)/g;
	if ( $d >= 2 ) {
		# Normalise multipart versions
		$s =~ s/(\.)(\d{1,3})/sprintf("$1%03d",$2)/eg;
	}
	$s =~ s/^(\d+)\.?//;
	my $l = $1 || 0;
	my @v = map {
		$_ . '0' x (3 - length $_)
	} $s =~ /(\d{1,3})\D?/g;
	$l = $l . '.' . join '', @v if @v;
	return $l + 0;
}

sub _cmp ($$) {
	_version($_[0]) <=> _version($_[1]);
}

# Cloned from Params::Util::_CLASS
sub _CLASS ($) {
	(
		defined $_[0]
		and
		! ref $_[0]
		and
		$_[0] =~ m/^[^\W\d]\w*(?:::\w+)*\z/s
	) ? $_[0] : undef;
}

1;

# Copyright 2008 - 2011 Adam Kennedy.


lib/Task/BeLike/HMA.pm
package Task::BeLike::HMA;

use 5.006001;

use strict;
use warnings;

our $VERSION = '0.09_01';

$VERSION = eval $VERSION;

__END__

=head1 NAME

Task::BeLike::HMA - Install modules HMA uses

=head1 VERSION

This is version 0.09_01 of Task::BeLike::HMA as of Aug 01, 2011.

=head1 SYNOPSIS

  perl Makefile.PL
  make

  # re-run Makefile.PL to see if all modules got installed
  perl Makefile.PL

  # make test and install if you want (not required)
  make test
  make install

=head1 DESCRIPTION

Task::BeLike::HMA is HMA's personal L<Task> to install the Perl
modules he likes to have available.

See C<Makefile.PL> for details.

=head1 DEPENDENCIES

This Task checks if recent versions of Perl toolchain modules are
available. If not, installation of C<Task::BeLike::HMA::Toolchain> is
suggested and nothing is installed.

The required minimum versions of toolchain modules are:

=over 4

=item * L<CPAN> 1.9301 (in Perl 5.8.9 core)

=item * L<Module::Build> 0.340201 (in 5.10.1 core)

=back

=head1 SEE ALSO

=over 4

=item * L<Task|http://search.cpan.org/perldoc?Task>

=item * L<Task::BeLike::HMA::Toolchain|https://github.com/hma/Task-BeLike-HMA-Toolchain>

=back

=head1 AUTHOR

Henning Manske <hma@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2010-2011 Henning Manske. All rights reserved.

This module is free software. You can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/>.

This module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


Makefile.PL
#
#  Task::BeLike::HMA - Makefile.PL
#
#  Copyright (c) 2010-2011 Henning Manske. All rights reserved.
#
#  This program is free software. You can redistribute it and/or modify it
#  under the terms of either: the GNU General Public License as published
#  by the Free Software Foundation; or the Artistic License.
#
#  See http://dev.perl.org/licenses/
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#

use 5.006001;

use strict;
use warnings;

BEGIN {
  # hack: load Module::CPANTS::Analyse before AutoInstall is loaded
  # TODO: bug report
  eval { require Module::CPANTS::Analyse };

  # Test::Tester warns on load if not loaded before Test::Builder
  eval { require Test::Tester };

  # prevent Carp::Always from installing its signal handlers
  eval { require Carp::Always };

  # force Module::AutoInstall to use CPAN.pm
  $ENV{PERL_AUTOINSTALL_PREFER_CPAN} = 1;
}

use inc::Module::Install;

# hack: install Module::Install from a Module::Install based Makefile.PL
# also, $VERSION of bundled inc/Module/Install.pm must be hacked to a lower
# version than required from the 'requires' command

use Module::AutoInstall;
shift @INC;



#
#  preliminary special setup for this Task
#

# check for recent versions of toolchain modules

unless ($ENV{PERL5_CPAN_IS_RUNNING}) {
  require CPAN;
  delete $ENV{PERL5_CPAN_IS_RUNNING};
  delete $ENV{PERL5_CPANPLUS_IS_RUNNING};
}

my $cpan   = eval { CPAN->VERSION } || 0;
my $mb     = eval { require Module::Build; Module::Build->VERSION } || 0;
my $lwp    = eval { require LWP; LWP->VERSION } || 0;
my $sqlite = eval { require CPAN::Config; $CPAN::Config->{use_sqlite} };
my $tymv   = eval { require Test::YAML::Meta::Version };

my $WANT_CPAN = '1.9301'; # version in Perl 5.8.9 core
my $WANT_MB   = '0.3603'; # 5.12.0

my $task = 'Task::BeLike::HMA::Toolchain 0.07_02';
my $url  = 'https://github.com/hma/Task-BeLike-HMA-Toolchain';

if (eval $cpan < $WANT_CPAN || eval $mb < $WANT_MB) {
  print "*** Toolchain too old.\n"
    . "    Consider installing $task first.\n"
    . "    See: $url\n";
  exit;
}

# suggest installation of Toolchain task

eval "use $task";
if ($@) {
  no warnings 'once';
  *MY::postamble = sub {
    print "\n*** Consider installing $task as well.\n"
      . "    See: $url\n\n";
    return "all ::\n"
      . "\t\$(NOECHO) \$(ECHO)\n"
      . "\t\$(NOECHO) \$(ECHO) \"*** Consider installing $task as well.\"\n"
      . "\t\$(NOECHO) \$(ECHO) \"    See: $url\"\n"
      . "\t\$(NOECHO) \$(ECHO)\n";
  }
}



#
#  module metadata
#

my $NAME = 'Task::BeLike::HMA';

(my $dist = $NAME) =~ s|::|-|g;
(my $path = $NAME) =~ s|::|/|g;

name       $dist;
license    'perl';
all_from   "lib/$path.pm";

bugtracker "https://github.com/hma/$dist/issues";
repository "https://github.com/hma/$dist";

#
#  notes
#
#  * autodie 2.10 t/caller.t fails if Perl < 5.8.4
#  * Module::Version requires autodie
#  * Perl::Critic 1.109 fails with '":color_severity" is not exported by
#    the Perl::Critic::Utils::Constants module' if Perl < 5.8.4
#  * Perl::MinimumVersion requires Perl::Critic
#  * Test::MinimumVersion requires Perl::MinimumVersion
#  * Test::CheckChanges 0.14 causes
#      1..1
#      # No tests run!
#    further on "make"
#      make: *** [installdeps] Error 255
#    TODO: bug report
#  * Module::Version 0.12 t/app.t fails if Getopt::Long < 2.36
#    TODO: bug report
#  * Test::Memory::Cycle 1.04 has unneccessary dependency PadWalker
#    TODO: bug report
#  * Data::Dump 1.18+ t/quote.t fails on 5.6.2, 5.8.0
#

requires 'Task::BeLike::HMA::Toolchain' => '0.07_02';

# Perl 5.10+ core pragmas

requires 'autodie'                      => '2.10'   if $] >= 5.008_004; # 5.13.3
requires 'parent'                       => '0.221'; # 5.10.1

# SSL support for LWP

if ($lwp < 6) {
  requires 'Crypt::SSLeay'              => '0.58';
}
else {
  requires 'LWP::Protocol::https'       => '6.02';
}

# common

requires 'Clone'                        => '0.30';
requires 'List::Util'                   => '1.23';  # 5.13.1
requires 'List::MoreUtils'              => '0.32';
requires 'Module::Pluggable'            => '3.9';   # 5.10.1
requires 'Readonly'                     => '1.03';
requires 'Readonly::XS'                 => '1.05'   if $] >= 5.008;
requires 'Task::Weaken'                 => '1.02';

# personal favourites

requires 'Exception::Class'             => '1.26'   if $] >= 5.008_001;
requires 'Package::Generator'           => '0.102'  if $] >= 5.008;
requires 'Sub::Exporter'                => '0.981';
requires 'Try::Tiny'                    => '0.07';

# testing

requires 'Test::Tester'                 => '0.107';
requires 'Test::NoWarnings'             => '1.02';
requires 'Test::Deep'                   => '0.108';
requires 'Test::Output'                 => '1.01';
requires 'Test::Exception'              => '0.31';
requires 'Test::Memory::Cycle'          => '1.04'   if $] >= 5.008_002;
requires 'Test::Requires'               => '0.04';
requires 'Test::Script'                 => '1.07'   if $] >= 5.008;
requires 'Test::Warn'                   => '0.23';

# personal favourites

requires 'aliased'                      => '0.30';
requires 'Carp::Always'                 => '0.06';
requires 'Carp::Clan'                   => '6.02';
requires 'Config::General'              => '2.50';
requires 'Data::Dump'                   => $] >= 5.008_001 ? '1.19' : '1.17';
requires 'Data::Hexify'                 => '1.00';
requires 'Data::Record';
requires 'File::Next'                   => '1.06';
requires 'JSON::XS'                     => '2.3'    if $] >= 5.008_002;
requires 'JSON'                         => '2.52';
requires 'Parse::RecDescent'            => '1.964';
requires 'Path::Class'                  => '0.24';
requires 'Text::CSV_XS'                 => '0.81';

# author / release testing

if ($] >= 5.008_001) {
  requires 'Pod::Simple'                => '3.16';  # 5.13.11
  requires 'Test::Pod'                  => '1.42';
  requires 'Test::Pod::No404s'          => '0.01';
}

requires 'Pod::Coverage'                => '0.19';
requires 'Test::Pod::Coverage'          => '1.08';
requires 'Test::CheckChanges'           => '0.14';
requires 'Test::CPAN::Meta'             => '0.17';
requires 'Test::DistManifest'           => '1.009';
requires 'Test::EOL'                    => '0.9'    if $] >= 5.008;
requires 'Test::HasVersion'             => '0.012';

if ($] >= 5.008_002 && ! $sqlite || $tymv) {
  #
  # CPAN::SQlite indexes only the latest release of a distribution (RT #63732)
  # Module::CPANTS::Analyse requires Test::YAML::Meta::Version
  # which is no longer contained in recent Test-YAML-Meta (RT #65903)
  # so if CPAN::SQlite is enabled and T::Y::M::Version is not installed
  # manual installation of T::Y::M::Version is required to enable
  # installation of Module::CPANTS::Analyse and Test::Kwalitee
  #
  # Test::YAML::Meta::Version is contained in
  # http://search.cpan.org/CPAN/authors/id/B/BA/BARBIE/Test-YAML-Meta-0.16.tar.gz
  #
  requires 'Module::CPANTS::Analyse'    => '0.85';
  requires 'Test::Kwalitee'             => '1.01';
}

if ($] >= 5.008_004) {
  my $perl_critic = eval {
    require List::MoreUtils;
    eval List::MoreUtils->VERSION <= 0.2703
  } ? '1.109' : '1.111';

  requires 'Perl::Critic'               => $perl_critic;
  requires 'Perl::MinimumVersion'       => '1.28';
  requires 'Test::MinimumVersion'       => '0.101080';
}

requires 'Test::Module::Used'           => '0.1.9'  if $] >= 5.008_003;
requires 'Test::NoTabs'                 => '1.0';
requires 'Test::Portability::Files'     => '0.05';
requires 'Test::Spelling'               => '0.14';
requires 'Devel::Cover'                 => '0.77';
requires 'Test::Strict'                 => '0.14'   if $] >= 5.008;
requires 'Test::Synopsis'               => '0.05'   if $] >= 5.008_001;

# admin

requires 'Getopt::Long'                 => '2.36';  # 5.9.5
requires 'Module::Version'              => '0.10'   if $] >= 5.008_004;
requires 'Module::Install'              => '1.01';
requires 'App::moduleshere'             => '0.07';
requires 'App::cpanminus'               => '1.4008' if $] >= 5.008;

if ($] >= 5.008_001) {
  requires 'local::lib'                 => '1.008002';
  requires 'App::cpanlistchanges'       => '0.05';
  requires 'App::cpanoutdated'          => '0.18';
  requires 'App::pmuninstall'           => '0.18';
}

if (index( $^X, 'perlbrew' ) == -1) {
  requires 'App::perlbrew'              => '0.27'   if $] >= 5.008;
  requires 'CPAN::Mini'                 => '1.111006';
}


build_requires 'Test::More';

clean_files "$dist-* Debian_CPANTS.txt";
makemaker_args dist => { COMPRESS => 'gzip -9f' };

auto_install;

WriteAll;
